Feature/wallet redesign take two

goranjovic opened the pull request:
fixes #6843 again

### Summary:
Wallet Send Redesign - Wallet flow

Designs: https://www.figma.com/file/XUehMnhyD1FGcWzvGz6SXqvh/Mobile-wallet?node-id=160%3A1707

Contains:

* Wallet Send flow (tokens, ether) as per designs.

Does not contain:

* dapp transactions redesign (still being redirected to the old wallet send flow)
* chat transactions redesign (also still being redirected to the old wallet send flow)
(the idea is that we redirect dapp and chat transactions to the new flow in smaller follow-up PRs
upon successful merge of this one. and all of them followed by a cleanup PR to remove the old wallet altogether)

Within Wallet flow omitted features are:

* max button
* "I know I'm on mainnet" when in dev mode

Other notes:

* Address entered by QR scanning is not displayed in checksummed format, but does pass the validation

**Functional**

* 1-1 chats (payments)
* wallet / transactions
* dapps / app browsing

status: ready

goranjovic committed:
wallet redesign take two - wallet flow only

goranjovic committed:
separated navigation stacks for different versions of send

goranjovic committed:
missing icons, gas callback

goranjovic committed:
using optimal gas when available

goranjovic committed:
disabled chat message

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 8acbcd91 | [#1](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-7916/1/) | 2019-04-04 14:39:28 | ~15 min | `macos` | [:package: dmg](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190404-142422-8acbcd-pr.dmg) |

annadanchenko commented:
@goranjovic can you provide link to wallet design covered here again, please? one from #6843 returns 404

goranjovic commented:
@annadanchenko right, the link is different now. here it is - https://www.figma.com/file/XUehMnhyD1FGcWzvGz6SXqvh/Mobile-wallet?node-id=160%3A1707

antdanchenko commented:
@goranjovic should I add tests from #7263 to this PR?

goranjovic commented:
@antdanchenko yes, but keep in mind that unlike the previous attempt this PR does not change the behavior of dapp and chat transactions

annadanchenko commented:
@goranjovic tests revealed an issue with sending tokens from chat: upon signing transaction with password there is error "java.lang.ClassCastException: com.facebook.react.bridge.ReadableNativeMap cannot be cast to java.lang.String" https://assets.saucelabs.com/jobs/8f9475db79a3474d880da6792948ee19/0022screenshot.png?auth=40330626bd8a86dc9579a0f723b243b1

Would expect that it can be sent

annadanchenko commented:
### issue 2
@goranjovic After sending ETHRo transaction from chat or from wallet to the contact there is no send message in the chat

goranjovic committed:
ignoring checksum warning on addresses scanned from qr

goranjovic commented:
@annadanchenko 
issue 1: looking into
issue 2: this was me, I disabled those messages. will re-enable

hesterbruikman commented:
@goranjovic eager to check this out @goranjovic ! Will look install and test the build over the weekend.

rasom commented on src/status_im/ui/screens/wallet/send/events.cljs:
```diff
@@ -202,8 +208,8 @@
      (fx/merge cofx
-               #(when send-command?
-                  (commands-sending/send % chat-id send-command? params))
+               #_#(when send-command?
```
Should this code be removed?

goranjovic replied:
No, it should just be re-enabled to send the chat message (Anna's point number 2)

goranjovic replied:
And I re-enabled it. Could you re-review?

rasom replied:
on it

annadanchenko commented:
@goranjovic can you look at problem detected by packages-check-bot, please? https://github.com/status-im/status-react/pull/7916/checks?check_run_id=94067962

goranjovic committed:
re-enabled chat messages

goranjovic commented:
@annadanchenko at the first sight seems unrelated to the PR. I just rebased and pushed the commit that re-enables the chat messages after wallet send. It should trigger a build so let's see what happens

annadanchenko commented:
### issue 3 Buttons are hidden behind keyboard on Send amount (Android only)
![Screenshot_20190408-123502](https://user-images.githubusercontent.com/6584810/55719102-aa48b380-59fd-11e9-823f-45601fd3027d.png)

Steps:
1. Wallet -> Send -> Contacts -> Select a contact -> check buttons at the bottom of screen

Android, real Galaxy S6, build 4

annadanchenko commented:
### issue 4 No "Done" button on iOS after sending tx. User needs to restart app to get back to Wallet
![18726](https://user-images.githubusercontent.com/6584810/55719486-a6696100-59fe-11e9-91d2-612771c97e2e.jpg)

Steps:
1. Wallet -> Send -> Contacts -> Select a contact -> type amount -> Next -> Confirm -> enter password -> Send

real iphone 7 plus, build 4

annadanchenko commented:
issue 5 Black text in Status bar (iOS and Android)

On all screens for send flow status bar has black text while it should be white


![Screenshot_20190408-130849](https://user-images.githubusercontent.com/6584810/55719883-a1f17800-59ff-11e9-84b8-4e9ab9b524de.png)

yenda commented on src/status_im/contact/db.cljs:
```diff
@@ -81,7 +81,7 @@
          (public-key->new-contact public-key))))
 
-(defn- contact-by-address [[_ contact] address]
+(defn- contact-by-address [contact address]
```
wouldn't the subscription `:contacts/contact-by-address` also be affected by this change?

goranjovic replied:
it does, but the only usage of that sub is being made unused by this PR and will be removed anyway

yenda replied:
so could you remove it rather than keep it since it won't work anymore?

goranjovic replied:
If I start removing code, it'd start domino effect and this PR size will grow to from 21 to 50-something in size :(
We'll have one cleanup PR after all three parts have been merged (wallet, chat and dapps)

yenda commented on src/status_im/ui/screens/wallet/choose_recipient/views.cljs:
```diff
@@ -46,9 +48,10 @@
-            dimensions        [:dimensions/window]
-            camera-flashlight [:wallet.send/camera-flashlight]]
+  (letsubs [{:keys [on-recipient]} [:get-screen-params :recipient-qr-code]
+            camera-flashlight      (reagent/atom :off) ;; values are :off and :on
```
how does one switch the camera on here?

goranjovic replied:
good catch, missed it

goranjovic replied:
added it back, there is a button in the toolbar that toggles the flashlight status

yenda commented on src/status_im/ui/screens/wallet/main/views.cljs:
```diff
@@ -68,7 +68,7 @@
     :accessibility-label :send-transaction-button
     :icon                :main-icons/send
-    :action              #(re-frame/dispatch [:navigate-to :wallet-send-transaction])}
+    :action              #(re-frame/dispatch [:navigate-to :wallet-send-transaction-new])}
```
can you namespace new screens?

goranjovic replied:
this name is temporary, it will be back to normal `:wallet-send-transaction` after dapp and chat transactions are also merged and old wallet flow is removed. 
Or did you mean, namespace them, in general, e.g. `:wallet/send-transaction..` etc?

goranjovic committed:
toggle flashlight

rasom commented on src/status_im/ui/screens/wallet/send/redesigned_views.cljs:
```diff
@@ -0,0 +1,98 @@
+
+;; SEND TRANSACTION FROM WALLET (CHAT)
+(defview send-transaction []
+  (letsubs [transaction    [:wallet.send/transaction]
```
It seems that `:wallet.send/transaction` subscription causes reloading of the screen right after rendering. In result text-input is focused twice and keyboard/view twitches.

goranjovic replied:
Good catch, I'll remove the subscription then. It's not even being used in the wallet flow

rasom replied:
> and keyboard/view twitches

at least on android real device

rasom commented on src/status_im/ui/screens/wallet/send/redesigned_views.cljs:
```diff
@@ -0,0 +1,98 @@
+;; MAIN SEND TRANSACTION VIEW
+(defn- send-transaction-view [_opts]
+  (reagent/create-class
+   {:reagent-render (fn [opts] [recipient/render-choose-recipient opts])}))
```
What's the point of creating class here, why `recipient/render-choose-recipient` can't be used directly?

goranjovic replied:
historical leftover, can remove

rasom replied:
that's not an important thing, it's just marked as a new code that's why I'm commenting

goranjovic replied:
yeah, because it's a new file

rasom commented on src/status_im/ui/screens/wallet/send/views/recipient.cljs:
```diff
@@ -0,0 +1,238 @@
+
+(defn update-recipient [chain transaction error-message value]
+  (if (ens/is-valid-eth-name? value)
+    (do (ens/get-addr (get ens/ens-registries chain)
```
`do` is not needed here

goranjovic replied:
historical leftover, can remove

rasom commented on src/status_im/ui/screens/wallet/send/events.cljs:
```diff
@@ -385,3 +391,109 @@
+  (if (ens/is-valid-eth-name? recipient)
+    (ens/get-addr (get ens/ens-registries chain)
+                  recipient
+                  #(callback %))
```
could be just `callback` probably

rasom commented on src/status_im/ui/screens/wallet/send/events.cljs:
```diff
@@ -385,3 +391,109 @@
+  (eth-name->address chain to
+                     (fn [to]
+                       (if (ethereum/address? to)
+                         (if (and (not qr) (not to-ens) (not (eip55/valid-address-checksum? to)))
```
`(not (or qr to-ens (eip55/valid-address-checksum? to)))` would have less `not`, probably less readable though, idk

goranjovic replied:
I too thought it would've been less readable even though it's more concise. The way it is now corresponds to how you'd explain it out loud in English, so I kept it like that

rasom replied:
understood

rasom commented on src/status_im/ui/screens/wallet/send/views/amount.cljs:
```diff
@@ -0,0 +1,314 @@
+
+(defn converted-currency-amount [{:keys [input-amount inverted]} token fiat-currency prices]
+  (when (valid-input-amount? input-amount)
+    (if-not inverted
```
it could be
```clj
(some-> (common/token->fiat-conversion prices token fiat-currency input-amount)
        (money/with-precision (if inverted 8 2)))
```

rasom commented on src/status_im/ui/screens/wallet/send/views/amount.cljs:
```diff
@@ -0,0 +1,314 @@
+    (when-let [amount-bn (if inverted
+                           (common/fiat->token-conversion prices token fiat-currency input-amount)
+                           (money/bignumber input-amount))]
+      amount-bn
```
leftover?

goranjovic replied:
probably

rasom commented on src/status_im/ui/screens/wallet/send/views/amount.cljs:
```diff
@@ -0,0 +1,314 @@
+    (update-input-errors token fiat-currency prices)))
+
+(defn render-choose-amount [{:keys [web3
+                                    chain
```
not used

rasom commented on src/status_im/ui/screens/wallet/send/views/amount.cljs:
```diff
@@ -0,0 +1,314 @@
+                                           :padding-vertical    7
+                                           :max-width           290
+                                           :text-align-vertical :center}}]
+                [react/text {:style {:color               (if (not (string/blank? input-amount))
```
not of great importance again, but we could avoid using `not` by swapping colors

rasom commented on src/status_im/ui/screens/wallet/send/views/amount.cljs:
```diff
@@ -0,0 +1,314 @@
+      (assoc state :error-message error-msg)
+      state)))
+
+(defn update-input-amount [state input-str token fiat-currency prices]
```
can we somehow split this function? a bit hard to read

goranjovic replied:
sure

rasom commented on src/status_im/ui/screens/wallet/send/views/common.cljs:
```diff
@@ -0,0 +1,453 @@
+  (some-> anim-ref (get signal) (apply args)))
+
+(defn slide-up-modal
+  "Creates a modal that slides up from the bottom of the screen and
```
I think we  have a bottom sheet component or something (for instance when we create a new chat), can it be reused here? Or this one is different?

goranjovic replied:
Yeah, I know. This code had originally been written before that component was introduced. The consensus we got on this is to merge it as it is and then refactor-replace this occurence with the new component. The reason is to reduce the scope creep of the pr.

rasom replied:
understood

goranjovic committed:
removed unused subs and references

goranjovic committed:
added changes requested in the review

goranjovic committed:
restored contact-by-address sub

goranjovic commented:
@yenda @rasom Updated, please re-review

goranjovic committed:
fixed token transactions in chat

annadanchenko commented:
### issue 6 Error in address if paste address shown on Receive screen and tap Next

Error is shown if valid address `0x8f4a98de7812cc0effbb2a643a056fc70dadcb35` is copied/pasted from Receive screen of the Status app to Address in "Send to" screen after tap on Next button, so can't send funds to this address (also checked with other address copied from Status app "0x2e1ce0f514387a188f4aeff4ceb6c2c0dea66ca7" - have same error)
![16993](https://user-images.githubusercontent.com/6584810/55799518-dd5c7700-5ad1-11e9-8a1b-4e28a84333f2.jpg)


Address copied from 
![7995](https://user-images.githubusercontent.com/6584810/55799553-f402ce00-5ad1-11e9-8a40-a5f050c29a40.jpg)

goranjovic commented:
@annadanchenko that one is a different issue  - https://github.com/status-im/status-react/issues/4959
It should be like that on develop as well right now

goranjovic commented:
The one thing I can do now is to disable EIP-55 checks as it's not implemented everywhere across the app.

annadanchenko commented:
> @annadanchenko that one is a different issue - #4959
> It should be like that on develop as well right now

Same address can be successfully used to send test ETH on today's nightly iOS build if copy paste it in Enter recipient address field. I believe we didn't implement checking for checksummed addresses yet

![11124](https://user-images.githubusercontent.com/6584810/55800981-4abdd700-5ad5-11e9-9937-fab05b0d2efb.jpg)

goranjovic commented:
@annadanchenko we did, but apparently it's broken on develop. Ok then, no harm will come if I disable it here as well.

antdanchenko commented:
@goranjovic I am going to push reworked tests, please mind my commit before doing force push

goranjovic commented:
@antdanchenko will do, thanks for the heads up

antdanchenko committed:
reworked wallet tests

goranjovic committed:
temporarily disabled eip55 validation

antdanchenko committed:
e2e tests fix

annadanchenko commented:
### issue 7 Zero USD network fee is shown for token transaction on Send amount screen and on Network fee settings.

Only for token transaction there is `~0 USD` shown as network fee. Note that for ETH transaction network fee is smaller but it's shown fine like `~0.008 USD`. 

![fee](https://user-images.githubusercontent.com/6584810/55826118-8ec9cf80-5b07-11e9-890d-aa53935efcd4.png)


On same screens for ETH transaction fee (even if it's smaller than fee for token tx) is shown fine, e.g.
![IMG_3629](https://user-images.githubusercontent.com/6584810/55826244-dd776980-5b07-11e9-81ba-15e1435f30bc.jpg)

annadanchenko commented:
### issue 8 Fee value is not updated after switching from Custom to Optimal

Video: https://drive.google.com/open?id=1S2PI5EMeVZSDQbkXYOFPY2txUl9aqIZd

Steps:
1. Wallet -> Send -> Contacts -> Select a contact 
2. type valid ETH amount (check default network fee shown)
3. tap on "network fee" to open Network fee settings
4. Tap on Custom  and increase gas price from 2 to 20000 Gwei
5. Tap Update. Check that network fee is changed on Send amount screen
6. Tap on network fee again
7. Tap on Optimal option and tap Update
8. Check value shown for network fee

annadanchenko commented:
Issue 9 Long amount is not moved to second line on Android 

Android 6.0.1, real Galaxy S6, build 12

Steps:

1. Wallet -> Send -> Contacts -> Select a contact
2. type valid ETH amount of `123,123456789012345` and check if end of the amount value is moved to the second line (As on design and as on iOS)

![Screenshot_20190409-214151](https://user-images.githubusercontent.com/6584810/55830296-e91b5e00-5b10-11e9-8e15-6ab01e2d2414.png)

annadanchenko commented:
### Issue 10 Amounts longer than 21 char can't be entered

Nothing happens if type amount with 21 chars and try to type extra number. E.g. user wants to send `12345,123456789012345678` SNT but can type only `12345,1234567890123`. We had such issue in the past: https://github.com/status-im/status-react/issues/4404

annadanchenko commented:
### Issue 11 No error is shown if enter ETH amount equal to wallet balance ( amount < ETH balance + fee). Would expect `not enough ETH for gas` error

User can proceed and sign transaction if there is not enough ETH to pay for gas. For example:
1. there is 0.002 ETH and user tries to send 0.002 ETH
2. there is no ETH and user tries to send some token transaction

I would expect same error as it's shown in 0.11.0 release (before redesign): not enough ETH for gas to be shown as soon as user types amount or after he taps on Next button. 

Steps:
0. You need to have some ETH in wallet
1. Wallet -> Send -> Contacts -> Select a contact
2. type exactly same amount of ETH that you have in Wallet (in my case it was 0.002 ETH)
3. Tap on Next and check if there is any error message shown

![Screenshot_20190409-221238](https://user-images.githubusercontent.com/6584810/55832236-6052f100-5b15-11e9-88af-4beece82bbdf.png)

annadanchenko commented:
@goranjovic can you add notes to the description on things not covered in PR but present in design, please? For example, "Max button" on Send amount screen; 0.3 min in Network fee settings; "I know that I'm on mainnet", etc. Or maybe design should be updated?

goranjovic commented:
@annadanchenko the first two are omitted. "Pending" label has been changed in the meantime and we should update that in designs.

goranjovic committed:
fixed gas cost in fiat for tokens

goranjovic committed:
reseting gas on choosing optimal

annadanchenko commented:
@dmitryn @goranjovic waiting for new build with keycard enabled

goranjovic committed:
enabled hardwallet

goranjovic commented:
@annadanchenko pushed keycard flag and fixes for issues 7, 8 and 10

goranjovic committed:
fixed validation of sufficient gas on top of amount

goranjovic commented:
and just fixed 11

annadanchenko commented:
### issue 12 Total in fiat value does not take into account network fee 

Expected: Total in fiat shows amount of funds to be sent + network fee. E.g. 0.02 USD + 0.095 USD. However, now only amount of funds to be sent is taken into account for Total, so only $ 0.02 is shown while it should be $0.115

iOS, build 12 and Android build 13

Steps:

1. You need to have some SNT in wallet
2. Wallet -> Send -> Contacts -> Select a contact
3. change asset to SNT. 
4. type SNT amount you have
5. Tap on Next and check fiat value shown under Total

![IMG_3644](https://user-images.githubusercontent.com/6584810/55883223-385f9e00-5ba6-11e9-8f33-43165b80d3c7.jpg)


See in designs that total is for amount + fee:

<img width="322" alt="Screen Shot 2019-04-10 at 15 29 47" src="https://user-images.githubusercontent.com/6584810/55883517-bf147b00-5ba6-11e9-8192-1edca631a894.png">

goranjovic committed:
enabled keycard in local dev

goranjovic committed:
fixed navigation after keycard signing of chat transactions

annadanchenko commented:
@goranjovic  Network fee is shown as zero on the "summary" send amount screen if send tokens, so issue 7 is not fully fixed https://github.com/status-im/status-react/pull/7916#issuecomment-481380038 on both Android and ios. It happens both for optimal and for custom options:

![86626](https://user-images.githubusercontent.com/6584810/55888331-77deb800-5baf-11e9-9936-20691d3ee966.jpg)
![68944](https://user-images.githubusercontent.com/6584810/55888409-95ac1d00-5baf-11e9-8ca9-b40c708d5687.jpg)

churik commented:
### issue 13 "Invalid address" is shown if typing valid ENS name and tap next
You can proceed if you tap "Next" twice
**Steps:**
- open Wallet
- tap on "Send"
- type valid ENS name (i.e. "nastya.stateofus.eth")
- tap on "Next"
**Expected result:** can proceed to next screen
**Actual result:** 
![Monosnap 2019-04-10 17-34-15](https://user-images.githubusercontent.com/4557972/55893017-b7a99d80-5bb7-11e9-81c8-e787a3621902.png)
To proceed need to tap on "Next" second time.
Issue is not reproducible if you paste ens name.
Can reproduce every time on IPhone (IOS 11.4.1), and saw only once on LG V20 (Android 8)

annadanchenko commented:
> @annadanchenko pushed keycard flag and fixes for issues 7, 8 and 10

@goranjovic 
- issue 7 is not fully fixed see https://github.com/status-im/status-react/pull/7916#issuecomment-481720029
- issue 8 is verified OK
- issue 10 " Amounts longer than 21 char can't be entered" still can reproduce with same steps on fresh build 15 Android and build 13 for iOS

churik commented:
### issue 14 Scanned address is not shown if you change it after  proceed to "Send Amount" and go back

**Steps:**
- open Wallet
- tap on "Send"
- paste valid address 
- tap on "Next"
- go back
- erase address 
- tap on Scan
- scan valid address
**Expected result:** address is scanned and shown
**Actual result:**  no address is shown
**Video:** https://take.ms/sjF9v

Serhy commented:
### Issue 15
Conversion `USD <-> ETH` (SNT/DAI/etc) does not convert crypto <-> fiat but what it does is 1) Flips currencies (that's correct) 2) then USD value taken as crypto and conversion (secondary value - represented as ETH) shows USD value.

<img width="279" alt="Screenshot 2019-04-10 at 20 01 36" src="https://user-images.githubusercontent.com/8749671/55898532-7a4f0b00-5bcb-11e9-9e91-d2560df20bef.png">

 To reproduce:
1. Wallet -> Send -> Contacts -> Select a contact
2. Tap `USD <-> ETH` button
3. Set 1 as USD value.
4. Verify there is 0.01823 is the value displayed as ETH value 

### Issue 16
The 5th issue above (Black text in Status bar) fixed but on iPhoneX there is a white background now.

<img width="547" alt="Screenshot 2019-04-10 at 20 00 03" src="https://user-images.githubusercontent.com/8749671/55898442-407e0480-5bcb-11e9-94e6-d8b871bd4ad6.png">

annadanchenko commented:
### 17 Send flow for keycard is not adjusted yet: `no account has been selected, please login` error after entering password and tap Send

![Screenshot_20190411-094949](https://user-images.githubusercontent.com/6584810/55940141-cc317880-5c3f-11e9-96fd-25f6be6b65fa.png)

goranjovic commented:
@annadanchenko I'm working on the keycard flow right now. Will push in the next half hour. Heads up, there is a bug when token transactions are signed with keycard. I just reproduced it in develop too, so will leave it in and push a separate fix for that. Dapp and eth transactions seem to be fine though.

goranjovic committed:
added keycard flow

goranjovic commented:
Rebased and enabled keycard flow. Another thing I noticed there is that there is no chat message after keycard transactions and history entry looks empty while pending. Going to go through the items mentioned in the PR comments and the call first and take another look at this then.

goranjovic committed:
removed eth usd button

goranjovic committed:
made network fee button in overview untappable

goranjovic committed:
moved signing phrase from sign popup to confirm popup

goranjovic committed:
fixed status-bar color

goranjovic committed:
added ens and address silent pre-validation

Signed-off-by: Jakub Sokołowski <jakub@status.im>

goranjovic commented:
Additions:

- Moved signing phrase from Sign bottom sheet to Confirm bottom sheet
- Made Network fee button untappable on Overview screen
- Removed ETH <-> USD button
- fixed status bar colors (issue 5 and also reported by Sergey)
- fixed quick typing ENS validation issue (reported by Tanya) by disabling the Next button until the input is valid, as suggested by Andrei

Notes:

- Network fee removal from Overview incidentally fixed the remainder of issue 7
- Can't reproduce issue 14 any more.
- Issue 12 is really the consequence of summed vs separate amounts. We need to make a decision on that.

Going to take a look at the screen size issues now (items 3 and 4)  and then revisit the keycard.

annadanchenko commented:
@jakubgs @pedro @goranjovic ios build 15 https://ci.status.im/job/status-react/job/prs/job/ios/view/change-requests/job/PR-7916/15/ failed with
```
16:29:16  + nix-shell /Users/jenkins/workspace/status-react_prs_ios_PR-7916/default.nix --argstr target-os ios --run 'bundle exec fastlane ios nightly'
16:29:20  bundler: failed to load command: fastlane (/Users/jenkins/.bundle/bin/fastlane)
16:29:20  NameError: uninitialized constant FastlaneCore::UpdateChecker
16:29:20    /Users/jenkins/.bundle/gems/fastlane-2.117.1/fastlane/lib/fastlane/cli_tools_distributor.rb:122:in `ensure in take_off'
16:29:20    /Users/jenkins/.bundle/gems/fastlane-2.117.1/fastlane/lib/fastlane/cli_tools_distributor.rb:122:in `take_off'
16:29:20    /Users/jenkins/.bundle/gems/fastlane-2.117.1/bin/fastlane:23:in `<top (required)>'
16:29:20    /Users/jenkins/.bundle/bin/fastlane:23:in `load'
16:29:20    /Users/jenkins/.bundle/bin/fastlane:23:in `<top (required)>'
```

same issue happened in other PR #7959 https://ci.status.im/job/status-react/job/prs/job/ios/view/change-requests/job/PR-7959/2/console

jakubgs commented:
I took the liberty of rebasing the branch. iOS build is fine now.

churik commented:
### issue 18 when scan QR code with address (format `ethereum:address?amount=amount`) - "Next" button is not active

**Steps:**
- tap on "Send"
- scan address with amount (i.e. `ethereum:0x4653a2cd80049e2734e2fd7364d0ed548c49c7b3?amount=0.001`)
Can't proceed - button "Next" is disabled
<img width="226" alt="sdfsdfdsfsdf" src="https://user-images.githubusercontent.com/4557972/56021036-c0f64f80-5d08-11e9-80e2-0af524b283f0.png">
Reproduced on both Android and IOS.

goranjovic commented:
@churik Can you paste the sample QR here too please?

goranjovic committed:
fixed validation of qr addresses

annadanchenko commented:
### Issue 19 No cursor in the amount field shown on Android

build 18, Pixel XL

![Screenshot_20190412-120232](https://user-images.githubusercontent.com/6584810/56029661-2ce1b380-5d1b-11e9-8687-adbe265348fd.png)

goranjovic committed:
fixed overlaps with navigation tabs

Serhy commented:
### Issue 16 above 
(https://github.com/status-im/status-react/pull/7916#issuecomment-481779430) became worse (26th Jenkins build). Now there is a white bar and white device stats (time, battery etc)

<img width="558" alt="Screenshot 2019-04-12 at 13 29 16" src="https://user-images.githubusercontent.com/8749671/56031023-fe69d580-5d26-11e9-98fc-5090e89536a8.png">

annadanchenko commented:
### Issue 20 Network fee is not visible because of overlap with warning message on summary screen
Android, pixel xl, build 18
![Screenshot_20190411-174431](https://user-images.githubusercontent.com/6584810/56032160-b1cfcb80-5d21-11e9-88eb-35c2069e7d35.png)

goranjovic commented:
re issue 19 - Can't reproduce issue 19 on my real phone (sony ericson xperia). I'm assuming it has something to do with device models and/or api version. I'm proposing that we bounty it.

re issue 20 - we can't really fix this without a design change. I'm proposing that we reduce the bottom sheet height, until the new design is ready. I'd rather not keep the PR pending until that happens though.

about to push the fix for bottom tabs overlap and ios status-bar (16). will ping when ready

goranjovic committed:
fixed confirm bottom sheet overlap

goranjovic commented:
Pushed:
- rebase
- fix for screen spacing near navigation menu (issues 3 and 4)
- fix for Confirm bottom sheet covering transaction overview screen- made the "background" screen scrollable as it was suggested on the call last week. Made no changes to fonts, margins and sizes as I thought I'd have to.

Please hold on with testing, still have to push the ios status-bar fix from my mac.

goranjovic committed:
status-bar on ios x is blue

goranjovic commented:
@annadanchenko Just pushed the status-bar on iphone x fix, it's building it.

annadanchenko commented:
made an overview on current issue states, so it's easier to find what can be tested/needs fixing: https://docs.google.com/spreadsheets/d/1zdBmKkO1OlkW9IGsqBQg_RS0Q6yj5aVGCVWvRpy2WkQ/edit?usp=sharing

annadanchenko commented:
@goranjovic here is a summary of not-fixed issues and questions:

Not fixed:
9  Long amount is not moved to second line on Android https://github.com/status-im/status-react/pull/7916#issuecomment-481407727
10  Amounts longer than 21 char can't be entered https://github.com/status-im/status-react/pull/7916#issuecomment-481411008

Questions:
18  when scan QR code with address (format ethereum:address?amount=amount) - "Next" button is not active https://github.com/status-im/status-react/pull/7916#issuecomment-482477572 If such QR code is scanned then should be amount populated with value or not? When I scan QR code from this issue then I can proceed with Next but amount is 0

Serhy commented:
iPhoneX issue (whitebar) https://github.com/status-im/status-react/pull/7916#issuecomment-481779430 is fixed with the latest build.

annadanchenko commented:
### Issue 21 Empty space between keyboard and Next button on iOS on Send amount screen.

iphone 7 plus
![19293](https://user-images.githubusercontent.com/6584810/56145859-69284480-5fa5-11e9-9934-919c26f4985b.jpg)

iphone 5s
![97466](https://user-images.githubusercontent.com/6584810/56145876-734a4300-5fa5-11e9-9524-8bddffc49a94.jpg)

annadanchenko commented:
### Issue 22 Empty space between keyboard and Next button on Android if open network fee or switch to other tab and back

Video: https://drive.google.com/open?id=1g9RjLMWI942ePCfzxMlkgVsOigRrWJ87

Steps:
1. Wallet -> Send -> Contacts -> select a contact
2. On Send amount screen: tap on chat icon shown in the bottom navigation 
3. when navigated to chats tap on Wallet to return back
4. tap on amount filed and type 0.01
5. Check position of the button Next 
6. Instead of moving between Chats and Wallet you can tap on network fee, tap Update and tap on amount

Galaxy S6, Android 6.0.1 and Pixel XL, Android 9, build 22
![Screenshot_20190415-174200](https://user-images.githubusercontent.com/6584810/56146799-32532e00-5fa7-11e9-87f7-21230dbcd3e0.png)

annadanchenko commented:
### Issue 23 Changes in network fee are not taken into account to check ETH for gas 

No error "Not ehough ETH for gas" is shown and user can proceed to sending tx if set custom network fee that is higher than available amount. Would expect error shown and Next button not enabled.

Steps:
1. Wallet -> Send -> Contacts -> select a contact
2. on Send amount screen: type amount that is less than you have in wallet, so Next button is enabled and there is no error shown
3. tap on Network fee button
4. Tap on Custom and set Gas price to very high number, e.g. 50000 (so amount + fee > available amount)
5. Tap Update
6. Check if there is any error shown and if Next is enabled

goranjovic commented:
re issues 21 and 22 - this is actually the result of my fix for the screen spacing/overlaps around the tab menu. I probably overdid that a bit and will tone it down. But either way, we need an actual design for that, not just an ad hoc fix. Let's discuss on the design review and follow up with a bounty.

re issue 23 - on it and going through the list

goranjovic commented:
Regarding QR with amount - amount should be pre-populated in that case. Looking at it as well.

annadanchenko commented:
### Issue 24 Custom fee is not reset to default Optimal after changing an asset

Old custom fee value remains after selecting another asset, e.g changing ETH to SNT. It's confusing as amount is reset to 0. I would expect that custom fee is also reset and is set to Optimal.

1. Wallet -> Send -> Contacts -> select a contact
2. on Send amount screen: type some ETH amount
3. tap on Network fee button
4. Tap on Custom and set Gas price to very high number, e.g. 50000 (it's not mandatory, just easier to see that value will remain)
5. Tap Update
6. on Send amount screen: change asset from ETH to SNT. Note that amount is set to 0 SNT
7. type some amount to get network fee button visible
8. check value shown for network fee (it should be default one, not so big as you set in step 4)
9. tap on Network fee button and check which option is selected (expected is Optimal)

goranjovic committed:
increased max length for amount field

annadanchenko commented:
### Issue 25 No validation for Gas price value

In old implementation Gas price value has some validation, like min 1 wei error is shown if enter 0 or  0,0000000001 or -1. Currently, for all these values no error is shown. @goranjovic maybe you can share some extra unit tests or validations from old implementation after fixing it, so we can know if any other e2e should be covering this.

Steps:
1. Wallet -> Send -> Contacts -> select a contact
2. on Send amount screen: type some ETH amount
3. tap on Network fee button
4. tap on Custom
5. for gas price enter  0 or  0,0000000001 or -1
6. Check if any error message is shown about invalid value for gas price

annadanchenko commented:
### Issue 26 modified ENS name gets wrong value if switch to Contacts and back

#### Case 1: Existing name was entered and deleted, than non-existing name was entered, switch to Contacts and back -> first existing name is shown and Next is disabled

Video: https://drive.google.com/open?id=1VeOTxLBeWqKXlIfYjJ6hK1ALZUZcwfE-
Steps: 
1. make sure you are on mainnet
2. Wallet -> Send -> type "anna.stateofus.eth". Note that it's an existing name, button Next is enabled
3. Delete entered value but tapping on delete button on keyboard (so it deletes chars one by one from the very end to the beginning)
4. Type non-existing name "wrong.stateofus.eth" (or some other value that does not exist). Note that there is error "Unknown ENS name" and button Next is disabled
5. tap on Contacts
6. tap on Address
7. check value shown for ENS name and state of the Next button. 

Expected: last entered value is shown, so it's "wrong.stateofus.eth" thus error is shown above and NExt is disabled
Currently: "anna.stateofus.eth" is shown, button Next is disabled


#### Case 2 (todo)

goranjovic committed:
validating against balance after gas settings change

annadanchenko commented:
### Issue 27 Can proceed to sending tx if enter non-existing ENS name by modifying existing one

#### Case 1: Enter shorter name than allowed (a.stateofus.eth)
1. make sure you are on mainnet
2. Wallet -> Send -> type "a.stateofus.eth".
3. Check if you can proceed to sending tx (would expect error and Next disabled because it's shorter that allowed to register)

#### Case 2 Edit existing name to get invalid one:
1. make sure you are on mainnet
2. Wallet -> Send -> type "anna.stateofus.eth". Note that it's an existing name, button Next is enabled
3. Delete "ann" from the value, so "a.stateofus.eth" remains
4. Check if there is any error shown (in my case there is "Unknown ENS name" error) and the state of the Next button (in my case it's Enabled and it's an issue).

Note that according to terms and conditions "a" is too short and it was not possible to register such name in stateofus.eth

annadanchenko commented:
### Issue 28 Small UI issues box

#### 28.1 Rename second screen with Send amount title to Overview as in designs
![Screenshot_20190416-114908](https://user-images.githubusercontent.com/6584810/56199709-bbb53f80-603d-11e9-8d8a-3ce7b53b0fe5.png)

#### 28.2 ETH instead of ETHro is shown in total on Ropsten
![Screenshot_20190416-115020](https://user-images.githubusercontent.com/6584810/56199838-f15a2880-603d-11e9-8e79-bf0e654cd802.png)

#### 28.3 Different number formats ~71,721.568 USD vs no comma in ~71721.568
![IMG_3662](https://user-images.githubusercontent.com/6584810/56200951-24052080-6040-11e9-809e-b98077b5ed4b.jpg)
![IMG_3663 (1)](https://user-images.githubusercontent.com/6584810/56201083-69295280-6040-11e9-9bd2-c30cf6e6e28f.jpg)


#### 28.4 Remove $ in front of amount (~$1.64 USD) for Sending so it's as in designs
![IMG_3663 (2)](https://user-images.githubusercontent.com/6584810/56201293-dc32c900-6040-11e9-8dac-a201d61a0273.jpg)


#### 28.5 Change error "Not enough ETH for gas" to "Not enough ETH for network fee"
There is now concept called "network fee" so we need to report issue with it by specifying this name instead of word Gas that is not  mentioned on the Send screen

goranjovic committed:
resetting gas settings on asset change

goranjovic committed:
reworked gas settings validation

oskarth commented:
@goranjovic @annadanchenko @guylouis what's the state of this PR?

goranjovic commented:
Closing as deprecated, to be eventually superseded by the new redesign and multiple wallets work.
